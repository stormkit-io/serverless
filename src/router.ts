import type { Method } from "./utils/filesys";
import path from "path";
import {
  matchPath as mp,
  walkTree,
  parseFileName,
  fileToRoute,
} from "./utils/filesys";

export const matchPath = (
  directory: string,
  requestPath: string,
  requestMethod: string
): string | undefined => {
  return mp(walkTree(directory), requestPath, requestMethod)?.rel;
};

interface GenerateRoutesOptions {
  serverSide: boolean;
}

interface Route {
  path: string;
  file: string;
  method?: Method;
}

export const generateRoutes = (
  directory: string,
  options?: GenerateRoutesOptions
): Route[] => {
  const tree = walkTree(directory);
  const routes: Route[] = [];

  tree.forEach((file) => {
    const route = fileToRoute(file.rel);
    const parsed = parseFileName(file.name);

    if (
      parsed.method === "all" ||
      parsed.method === "get" ||
      options?.serverSide === true
    ) {
      const r: Route = {
        file: path.join("/", file.rel),
        path: route,
      };

      if (options?.serverSide === true) {
        r.method = parsed.method;
      }

      routes.push(r);
    }
  });

  return routes;
};
