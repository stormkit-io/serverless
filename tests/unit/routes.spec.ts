import path from "path";
import { generateRoutes } from "~/router";

describe("~/router", () => {
  const apiFolder = path.join(__dirname, "../../api");

  test("#generateRoutes - clientSide", () => {
    expect(generateRoutes(apiFolder)).toMatchSnapshot();
  });

  test("#generateRoutes - serverSide", () => {
    expect(generateRoutes(apiFolder, { serverSide: true })).toMatchSnapshot();
  });
});
