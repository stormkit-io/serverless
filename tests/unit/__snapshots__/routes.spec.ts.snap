// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`~/router #generateRoutes - clientSide 1`] = `
Array [
  Object {
    "file": "/404.ts",
    "path": "/404",
  },
  Object {
    "file": "/README.md",
    "path": "/README",
  },
  Object {
    "file": "/index.ts",
    "path": "/",
  },
  Object {
    "file": "/blog/about-me.ts",
    "path": "/blog/about-me",
  },
  Object {
    "file": "/blog/[name]/index.ts",
    "path": "/blog/:name",
  },
  Object {
    "file": "/blog/[name]/[date]/index.ts",
    "path": "/blog/:name/:date",
  },
  Object {
    "file": "/blog/[name]/[date]/list.ts",
    "path": "/blog/:name/:date/list",
  },
  Object {
    "file": "/user/index.ts",
    "path": "/user",
  },
  Object {
    "file": "/user/token.ts",
    "path": "/user/token",
  },
  Object {
    "file": "/user/[id]/index.get.ts",
    "path": "/user/:id",
  },
]
`;

exports[`~/router #generateRoutes - serverSide 1`] = `
Array [
  Object {
    "file": "/404.ts",
    "method": "all",
    "path": "/404",
  },
  Object {
    "file": "/README.md",
    "method": "all",
    "path": "/README",
  },
  Object {
    "file": "/index.head.ts",
    "method": "head",
    "path": "/",
  },
  Object {
    "file": "/index.put.ts",
    "method": "put",
    "path": "/",
  },
  Object {
    "file": "/index.ts",
    "method": "all",
    "path": "/",
  },
  Object {
    "file": "/blog/about-me.ts",
    "method": "all",
    "path": "/blog/about-me",
  },
  Object {
    "file": "/blog/[name]/index.ts",
    "method": "all",
    "path": "/blog/:name",
  },
  Object {
    "file": "/blog/[name]/[date]/index.ts",
    "method": "all",
    "path": "/blog/:name/:date",
  },
  Object {
    "file": "/blog/[name]/[date]/list.ts",
    "method": "all",
    "path": "/blog/:name/:date/list",
  },
  Object {
    "file": "/user/index.ts",
    "method": "all",
    "path": "/user",
  },
  Object {
    "file": "/user/token.ts",
    "method": "all",
    "path": "/user/token",
  },
  Object {
    "file": "/user/[id]/index.get.ts",
    "method": "get",
    "path": "/user/:id",
  },
  Object {
    "file": "/user/[id]/index.post.ts",
    "method": "post",
    "path": "/user/:id",
  },
]
`;
